Switch to flexbox

simonsmith opened the pull request:
This refactors the grid to use flexbox. The goal here was to keep the API the same so that it could be dropped into an existing site with the SUIT grid and everything work. This includes working with the other utilities where possible.

Here are the tests with the flexbox version - https://dl.dropboxusercontent.com/u/1709558/suitcss-grid/index.html I've also added a couple new bits like `fix` and `equalHeight`. I was able to get the existing tests to match without making any changes to the markup, which was a plus. 

This does depend on some changes that I've [made to utils-size](https://github.com/simonsmith/utils-size) as well. Namely losing the various `sizeFit` and `sizeFill` utils and adding `flex-basis` alongside `width`. The offset utils seem to work with no changes required.

A further enhancement to this will be to create a utils-flexbox (https://github.com/suitcss/suit/issues/113) which would allow even more control over the grid with things like `order` and `align-content` etc.

I'd suggest that makes grid `3.0.0` and once users are ready to support flexbox they should be able to update their package and expect everything to work as normal.

**Depends on** https://github.com/suitcss/utils-size/pull/29 

Fix #33 
Fix #35 
Fix #34 
Fix #15

@timkelty @giuseppeg

giuseppeg commented on README.md:
```diff
@@ -33,22 +34,35 @@ Read more about [SUIT CSS](https://github.com/suitcss/suit/).
 * `Grid--alignRight`: right-align all child `Grid-cell`
 * `Grid--alignMiddle`: middle-align all child `Grid-cell`
 * `Grid--alignBottom`: bottom-align all child `Grid-cell`
+* `Grid--fit`: evenly distribute space amongst all child `Grid-cell`
```
Maybe we can use the same api from `Arrange` and call this `Grid--equal` although `fit` sounds better probably :)

simonsmith replied:
Yeah, I think with equalHeight existing I wondered if equal alone is confusing. Copied Philip Waltons grid here on the end. Open to suggestions of course

giuseppeg replied:
Agreed :) `fit` is good!

giuseppeg commented on test/index.html:
```diff
@@ -5,11 +5,12 @@
-    background: rgb(100, 150, 50);
-    background: rgb(100, 150, 50);
-    border: 1px solid black;
+    background: rgba(100, 150, 50, .75);
```
Any particular reason why you tweaked the styles for `.fixture-Box`?

simonsmith replied:
The main reason was to copy the nesting examples here http://philipwalton.github.io/solved-by-flexbox/demos/grids/ but I changed my mind. I can revert this.

simonsmith replied:
The colour is much nicer on the eyes though :)

giuseppeg replied:
Should we adopt this across all the pkgs? Maybe we should move the fixture-Box to components-test

simonsmith replied:
Yeah, could be an idea. I think it's inline in all examples at the moment

giuseppeg commented on test/index.html:
```diff
@@ -5,11 +5,12 @@
+    background: rgba(100, 150, 50, .75);
+    border: 1px solid #888;
+    font-family: sans-serif;
     margin-bottom: 10px;
```
We should remove this margin it can be misleading.
People may make the assumption that wrapping cells with gutters also get vertical spacing.

By the way, maybe we should also add vertical spacing to `Grid--withGutters`.

simonsmith replied:
Yeah not sure on spacing as part of the grid. How do other ones handle this?

I think this raises a wider issue I faced which is lack of a proper demo page for users. Like http://flexboxgrid.com/

simonsmith replied:
So on reflection I don't think the grid should introduce any vertical spacing. It would break sites that used it for a content/sidebar layout if the bottom of their cells suddenly gained 20px. I've managed to drop this updated Grid into my blog and it all renders the same without any changes so i think that's a good goal to keep.

giuseppeg replied:
I think that it is a "bad" idea in terms of backwards compatibility too.
For the record I didn't think of adding `margin-bottom` but `margin-top` starting from the second "row" a bit like [this](http://www.yelp.com/styleguide#section-arrange) (scroll to the bottom of the section/see arrange--wrap)

simonsmith replied:
That seems to add margin to the top row too?

I see your idea though. Perhaps we could make it an opt-in modifier with a custom property for the width.

giuseppeg replied:
There is a -something on the container (although this workaround can have unexpected results sometimes).

Anyway I am sold, let's postpone this for when someone will ask about it :)

giuseppeg commented:
Looks good to me! Good job ;)

Wouldn't it be helpful to keep `u-sizeFill` to get a "expand to fill the remaining space" behaviour?
I mean it should already [work](http://codepen.io/giuseppegurgone/pen/zvaKoY) but for some reason it doesn't in your tests.

simonsmith commented:
Good point about `sizeFill`. After playing around I think we could do this:

``` html
    <div class="Grid">
      <div class="Grid-cell u-size1of4">
        <div class="fixture-Box">1</div>
      </div>
      <div class="Grid-cell u-sizeFill">
        <div class="fixture-Box">2</div>
      </div>
    </div>
```

``` css
.u-sizeFill {
  flex: 1;
}
```

<img width="957" alt="screen shot 2015-10-30 at 09 13 28" src="https://cloud.githubusercontent.com/assets/360703/10842078/ba787ba0-7ee6-11e5-80a6-69cfa3c08e21.png">

When combined with a proper `utils-flexbox` I expect you could do all the things that Arrange allows.

`.u-sizeFull` might still be useful too, so maybe keep those but ditch the other float based utils?

giuseppeg commented:
<blockquote>.u-sizeFull might still be useful too, so maybe keep those but ditch the other float based utils?</blockquote>
Yep sounds good to me!

simonsmith commented:
* Reverted my test colour changes (wait till we decide on any changes across all packages)
* Added a new test (1.6) to show usage of `u-sizeFill` as mentioned previously.

Ran the test page through Browserstack and it seems like it might need some work with IE10/11 (who'd have thought?)

https://www.dropbox.com/sh/rxd28o2ex2kgd13/AACSMioDEhbB-XRoIYn3HI5Ta?lst

simonsmith commented:
Tests passing as expected:

* Chrome - http://cl.ly/dgGC (expected result)
* IE10 - http://cl.ly/dfbD
* IE11 - http://cl.ly/dfqQ
* Safari - http://cl.ly/dfcc
* iOS - http://cl.ly/dgU0
* Firefox - http://cl.ly/dgYu

simonsmith commented:
Also wondering if current Grid should go to `1.0.0` and then such a big change as flexbox warrants a `2.0`? After all, the current Grid is used in production in several places and has been stable for a long time, so should have been `1.0` ages ago in my opinion.

giuseppeg commented on lib/grid.css:
```diff
@@ -46,31 +42,46 @@
+ */
+
+.Grid--fit > .Grid-cell {
+  flex: 1 1 0%; /* 1 */
```
can the `0` be unitless here?

simonsmith replied:
It's recommended not to because of IE10/11 - https://github.com/philipwalton/flexbugs#4-flex-shorthand-declarations-with-unitless-flex-basis-values-are-ignored

simonsmith replied:
Perhaps I should add that to the comments?

simonsmith replied:
Updated

simonsmith commented:
If no further issues I will merge this and the [utils-size](https://github.com/suitcss/utils-size/pull/29) changes
